frontend/server/handlers/artifacts.ts

```diff
@@ -74,6 +76,75 @@ export interface GCSProviderInfo {
   };
 }
 
+/**
+ * Returns an authorization middleware for artifact endpoints.
+ * This middleware validates that the user has permission to access artifacts
+ * in the specified namespace, preventing unauthorized cross-namespace access.
+ *
+ * Security: This addresses CVE-like vulnerability where namespace parameter
+ * could be manipulated to access artifacts from other namespaces.
+ * See: https://github.com/kubeflow/pipelines/issues/9889
+ *
+ * @param authorizeFn The authorization function to validate permissions
+ * @param authEnabled Whether authorization is enabled
+ * @param kubeflowUserIdHeader The header name containing the user identity
+ */
+export function getArtifactsAuthMiddleware(
+  authorizeFn: AuthorizeFn,
+  authEnabled: boolean,
+  kubeflowUserIdHeader: string,
+): Handler {
+  return async (req: Request, res: Response, next: NextFunction) => {
+    if (!authEnabled) {
+      return next();
+    }
+
+    const namespace = req.query.namespace as string | undefined;
+
+    if (!namespace) {
+      console.warn(
+        `[SECURITY] Artifact access denied: missing namespace parameter. ` +
```

Comment:
Only in some special non-default configurations if you enable the dangerous and overhead artifact proxy there will be the namespace parameter. Just that you are aware, that is is just one option.